Aborting ArrayIndexOutOfBoundsException while indexing embedded virtual disk segments or huge files

@gabrieldmf sent this to me:
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -65536 out of bounds for length 71428
	at org.apache.lucene.index.TermsHashPerField.writeByte(TermsHashPerField.java:219) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.TermsHashPerField.writeVInt(TermsHashPerField.java:241) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.FreqProxTermsWriterPerField.writeProx(FreqProxTermsWriterPerField.java:86) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.FreqProxTermsWriterPerField.newTerm(FreqProxTermsWriterPerField.java:127) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.TermsHashPerField.initStreamSlices(TermsHashPerField.java:175) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:198) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.IndexingChain$PerField.invert(IndexingChain.java:1224) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.IndexingChain.processField(IndexingChain.java:729) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.IndexingChain.processDocument(IndexingChain.java:620) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:241) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:432) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1532) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at org.apache.lucene.index.IndexWriter.addDocuments(IndexWriter.java:1503) ~[lucene-core-9.2.0.jar:9.2.0 ba8c3a806ada3d7b3c34d408e449a92376a8481b - romseygeek - 2022-05-19 15:10:13]
	at iped.engine.task.index.IndexTask.process(IndexTask.java:148) ~[iped-engine-4.0.2.jar:?]
```
It was processing an embedded 64GB flat vmdk disk partially encrypted by some ransomware, but it should work. Seems a Lucene issue to me, so I opened:
https://issues.apache.org/jira/browse/LUCENE-10681

--- a/iped-app/resources/config/IPEDConfig.txt
+++ b/iped-app/resources/config/IPEDConfig.txt
@@ -81,12 +81,6 @@ enableGraphGeneration = true
 # May cause loss of hits surrounded by "random" content.
 entropyTest = true
 
-# Large binary files or those without a specific parser will be split before being processed. 
-# This makes hit/regex highlighting in large files (pagefile, VSS, unallocated...) a lot faster
-# and allows exporting relevant data chunks to reports without exporting the whole file.
-# The original file is kept in case, although it is not processed by most modules.
-enableSplitLargeBinary = true
-
 # Indexes files contents. If disabled, indexes only the properties of files.
 indexFileContents = true
 

--- a/iped-engine/src/main/java/iped/engine/config/SplitLargeBinaryConfig.java
+++ b/iped-engine/src/main/java/iped/engine/config/SplitLargeBinaryConfig.java
@@ -1,15 +1,18 @@
 package iped.engine.config;
 
+import java.io.IOException;
+import java.nio.file.DirectoryStream.Filter;
+import java.nio.file.Path;
+
 import iped.utils.UTF8Properties;
 
-public class SplitLargeBinaryConfig extends AbstractTaskPropertiesConfig {
+public class SplitLargeBinaryConfig extends AbstractPropertiesConfigurable {
 
     /**
      * 
      */
     private static final long serialVersionUID = 1L;
 
-    private static final String ENABLE_PARAM = "enableSplitLargeBinary";
     private static final String CONF_FILE = "SplitLargeBinaryConfig.txt";
 
     private long minItemSizeToFragment = 100 * 1024 * 1024;
@@ -26,16 +29,6 @@ public class SplitLargeBinaryConfig extends AbstractTaskPropertiesConfig {
 
     public int getFragmentOverlapSize() {
         return fragmentOverlapSize;
-    }
-
-    @Override
-    public String getTaskEnableProperty() {
-        return ENABLE_PARAM;
-    }
-
-    @Override
-    public String getTaskConfigFileName() {
-        return CONF_FILE;
     }
 
     @Override
@@ -56,6 +49,16 @@ public class SplitLargeBinaryConfig extends AbstractTaskPropertiesConfig {
             fragmentOverlapSize = Integer.valueOf(value.trim());
         }
 
+    }
+
+    @Override
+    public Filter<Path> getResourceLookupFilter() {
+        return new Filter<Path>() {
+            @Override
+            public boolean accept(Path entry) throws IOException {
+                return entry.endsWith(CONF_FILE);
+            }
+        };
     }
 
 }

--- a/iped-engine/src/main/java/iped/engine/task/FragmentLargeBinaryTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/FragmentLargeBinaryTask.java
@@ -12,7 +12,6 @@ import iped.engine.datasource.SleuthkitReader;
 import iped.engine.task.carver.BaseCarveTask;
 import iped.engine.util.TextCache;
 import iped.parsers.standard.StandardParser;
-import iped.parsers.util.MetadataUtil;
 
 /**
  * Breaks large binary files (indexed by strings) into smaller pieces to be
@@ -24,7 +23,7 @@ import iped.parsers.util.MetadataUtil;
 public class FragmentLargeBinaryTask extends BaseCarveTask {
 
     // workaround for https://github.com/sepinf-inc/IPED/issues/1281
-    private static final long MIN_SPLIT_SIZE_IF_XHTML_OR_ERROR = 1 << 30;
+    private static final int TEXT_SPLIT_SIZE = 1 << 30;
 
     private SplitLargeBinaryConfig splitConfig;
     private StandardParser autoParser;
@@ -51,21 +50,16 @@ public class FragmentLargeBinaryTask extends BaseCarveTask {
         return !caseData.isIpedReport() && splitConfig.isEnabled();
     }
 
-    public static boolean isXHtmlToSplit(IItem evidence) {
-        return evidence.getLength() != null && evidence.getLength() >= MIN_SPLIT_SIZE_IF_XHTML_OR_ERROR
-                && (MetadataUtil.isHtmlMediaType(evidence.getMediaType())
-                        || MetadataUtil.isHtmlSubType(evidence.getMediaType()));
-    }
-
     @Override
     protected void process(IItem evidence) throws Exception {
         
         boolean hasSpecificParser = ParsingTask.hasSpecificParser(autoParser, evidence);
         boolean hadParserException = Boolean.valueOf(evidence.getMetadata().get(StandardParser.PARSER_EXCEPTION));
 
+        TextCache textCache = ((Item) evidence).getTextCache();
+
         if (evidence.getLength() != null && evidence.getLength() >= splitConfig.getMinItemSizeToFragment()
-                && (evidence.isTimedOut() || isXHtmlToSplit(evidence)
-                        || (hasSpecificParser && hadParserException && evidence.getLength() >= MIN_SPLIT_SIZE_IF_XHTML_OR_ERROR)
+                && (evidence.isTimedOut()
                         || (!hasSpecificParser && (!EmbeddedDiskProcessTask.isSupported(evidence)
                                 || !EmbeddedDiskProcessTask.isFirstOrUniqueImagePart(evidence) || hadParserException)))
                 && evidence.getInputStreamFactory() != null
@@ -76,19 +70,32 @@ public class FragmentLargeBinaryTask extends BaseCarveTask {
             int overlap = splitConfig.getFragmentOverlapSize();
             for (long offset = 0; offset < evidence.getLength(); offset += fragSize - overlap) {
                 long len = offset + fragSize < evidence.getLength() ? fragSize : evidence.getLength() - offset;
-                this.addFragmentFile(evidence, offset, len, fragNum++);
+                this.addFragmentFile(evidence, offset, len, fragNum++, null);
                 if (Thread.currentThread().isInterrupted())
                     return;
             }
 
             // set an empty text in parent
-            TextCache textCache = new TextCache();
-            ((Item) evidence).setParsedTextCache(textCache);
+            ((Item) evidence).setParsedTextCache(new TextCache());
+
+        } else if (textCache != null && textCache.getSize() > TEXT_SPLIT_SIZE) {
+            int fragNum = 0;
+            long totalTextSize = textCache.getSize();
+            for (long textOffset = 0; textOffset < totalTextSize; textOffset += TEXT_SPLIT_SIZE - splitConfig.getFragmentOverlapSize()) {
+                TextCache textCacheChunk = textCache.clone();
+                int textSize = textOffset + TEXT_SPLIT_SIZE < totalTextSize ? TEXT_SPLIT_SIZE : (int) (totalTextSize - textOffset);
+                textCacheChunk.setTextBounds(textOffset, textSize);
+                this.addFragmentFile(evidence, 0, evidence.getLength(), fragNum++, textCacheChunk);
+            }
+            if (fragNum > 0) {
+                // set an empty text in parent
+                ((Item) evidence).setParsedTextCache(new TextCache());
+            }
         }
 
     }
 
-    private void addFragmentFile(IItem parentEvidence, long off, long len, int fragNum) {
+    private void addFragmentFile(IItem parentEvidence, long off, long len, int fragNum, TextCache textCache) {
         String name = parentEvidence.getName() + "_" + fragNum; //$NON-NLS-1$
         Item fragFile = getOffsetFile(parentEvidence, off, len, name, parentEvidence.getMediaType());
         configureOffsetItem(parentEvidence, fragFile, off);
@@ -100,6 +107,13 @@ public class FragmentLargeBinaryTask extends BaseCarveTask {
         fragFile.setModificationDate(parentEvidence.getModDate());
         fragFile.setChangeDate(parentEvidence.getChangeDate());
         fragFile.setExtraAttribute(FILE_FRAGMENT, true);
+        if (textCache != null) {
+            fragFile.setParsedTextCache(textCache);
+        }
+        // avoid computing hash for fragments again
+        if (parentEvidence.getHash() != null && !parentEvidence.getHash().isEmpty()) {
+            fragFile.setHash("");
+        }
         addOffsetFile(fragFile, parentEvidence);
     }
 

--- a/iped-engine/src/main/java/iped/engine/task/FragmentLargeBinaryTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/FragmentLargeBinaryTask.java
@@ -47,7 +47,7 @@ public class FragmentLargeBinaryTask extends BaseCarveTask {
 
     @Override
     public boolean isEnabled() {
-        return !caseData.isIpedReport() && splitConfig.isEnabled();
+        return !caseData.isIpedReport();
     }
 
     @Override

--- a/iped-engine/src/main/java/iped/engine/task/ParsingTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/ParsingTask.java
@@ -316,7 +316,7 @@ public class ParsingTask extends ThumbTask implements EmbeddedDocumentExtractor 
                 .findObject(SplitLargeBinaryConfig.class);
         if (((Item) evidence).getTextCache() == null
                 && ((evidence.getLength() == null || !splitConfig.isEnabled() || evidence.getLength() < splitConfig.getMinItemSizeToFragment())
-                || (StandardParser.isSpecificParser(parser) && !FragmentLargeBinaryTask.isXHtmlToSplit(evidence)))) {
+                || StandardParser.isSpecificParser(parser))) {
             try {
                 depth++;
                 ParsingTask task = new ParsingTask(worker, autoParser);

--- a/iped-engine/src/main/java/iped/engine/util/TextCache.java
+++ b/iped-engine/src/main/java/iped/engine/util/TextCache.java
@@ -13,14 +13,16 @@ import java.io.Writer;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystemException;
 import java.nio.file.Files;
+import java.util.concurrent.atomic.AtomicInteger;
 
+import org.apache.commons.io.input.BoundedReader;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import iped.data.IItem;
 import iped.utils.IOUtil;
 
-public class TextCache implements Closeable {
+public class TextCache implements Closeable, Cloneable {
 
     private static Logger logger = LoggerFactory.getLogger(TextCache.class);
 
@@ -32,6 +34,8 @@ public class TextCache implements Closeable {
     private Writer writer;
     private long size = 0;
     private boolean diskCacheEnabled = true;
+    private long offset = -1;
+    private AtomicInteger refCount = new AtomicInteger();
 
     public void setSourceItem(IItem sourceItem) {
         this.sourceItem = sourceItem;
@@ -100,17 +104,36 @@ public class TextCache implements Closeable {
             }
         }
 
-        if (reader != null)
+        if (reader != null) {
+            if (offset > -1) {
+                long skipped = 0;
+                while (skipped < offset) {
+                    skipped += reader.skip(offset - skipped);
+                }
+                reader = new BoundedReader(reader, (int) size);
+            }
             return new KnownSizeReader(reader);
+        }
 
         return null;
+    }
+
+    public void setTextBounds(long offset, int size) {
+        if (offset < 0 || size < 0) {
+            throw new IllegalArgumentException("Both offset & size must be non negative.");
+        }
+        if (offset + size > this.size) {
+            throw new IllegalArgumentException("offset + size must be less than or equal to original text size.");
+        }
+        this.offset = offset;
+        this.size = size;
     }
 
     @Override
     public void close() throws IOException {
         if (writer != null)
             writer.close();
-        if (tmp != null)
+        if (tmp != null && refCount.decrementAndGet() == 0)
             tmp.delete();
     }
 
@@ -136,6 +159,22 @@ public class TextCache implements Closeable {
             return size;
         }
 
+    }
+
+    @Override
+    public TextCache clone() {
+        TextCache o = new TextCache();
+        o.sourceItem = sourceItem;
+        o.sb = sb;
+        o.tmp = tmp;
+        // we just use clone for reading for now
+        // o.writer = writer;
+        o.size = size;
+        o.diskCacheEnabled = diskCacheEnabled;
+        o.offset = offset;
+        o.refCount = refCount;
+        refCount.incrementAndGet();
+        return o;
     }
 
 }

--- a/iped-engine/src/main/java/iped/engine/util/TextCache.java
+++ b/iped-engine/src/main/java/iped/engine/util/TextCache.java
@@ -1,12 +1,15 @@
 package iped.engine.util;
 
+import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.Closeable;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.io.OutputStreamWriter;
+import java.io.RandomAccessFile;
 import java.io.Reader;
 import java.io.StringReader;
 import java.io.Writer;
@@ -16,6 +19,7 @@ import java.nio.file.Files;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import org.apache.commons.io.input.BoundedReader;
+import org.apache.commons.io.input.RandomAccessFileInputStream;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -94,7 +98,14 @@ public class TextCache implements Closeable, Cloneable {
 
         if (tmp != null) {
             try {
-                reader = Files.newBufferedReader(tmp.toPath());
+                if (offset == -1) {
+                    reader = Files.newBufferedReader(tmp.toPath());
+                } else {
+                    RandomAccessFile raf = new RandomAccessFile(tmp, "r");
+                    raf.seek(offset);
+                    reader = new InputStreamReader(new RandomAccessFileInputStream(raf, true), StandardCharsets.UTF_8);
+                    reader = new BufferedReader(new BoundedReader(reader, (int) size));
+                }
 
             } catch (FileSystemException | FileNotFoundException e) {
                 logger.error("Error reading extracted text file{}, maybe your antivirus blocked or deleted it? {}",
@@ -105,13 +116,6 @@ public class TextCache implements Closeable, Cloneable {
         }
 
         if (reader != null) {
-            if (offset > -1) {
-                long skipped = 0;
-                while (skipped < offset) {
-                    skipped += reader.skip(offset - skipped);
-                }
-                reader = new BoundedReader(reader, (int) size);
-            }
             return new KnownSizeReader(reader);
         }
 

